### DISCUSSION TITLE
offline development

### DISCUSSION THREAD

**Discussion by mwinter80** [raspberrypi/picamera2#765] [Q&A]
Hello, is it possible to run picamera2 with dummy data for development offline or on a non-pi computer? For example, using a video as opposed to the feed from the camera.

Thanks in advance!

**Comment by davidplowman** [comment]
I'm afraid I don't know of any direct way to do that. I guess you could do something like
```
import cv2
from picamera2 import Picamera2

def images_from_file(filename):
    cap = cv2.VideoCapture(filename)
    while True:
        _, frame = cap.read()
        yield frame

def images_from_camera():
    picam2 = Picamera2()
    picam2.start()
    while True:
        yield picam2.capture_array()
```
which would allow this
```
for frame in images_from_camera():  # or for development: for frame in images_fom_file(filename):
    # do something with frame
```
Picamera2 is unlikely to work on any non-Pi computers. In theory there's a chance if the platform has an implementation of libcamera (which may be the case on some Linux systems), though you might also find differences in what is supported if your use case is not reasonably "simple".

**Comment by mwinter80** [comment]
I was thinking something along those lines, but is it possible to pipe that back in to QGLPicamera2 if you are working on an app?

**Comment by davidplowman** [comment]
Indeed, that's sounding tricky. The camera buffers are dmabufs that can be shared directly with the 3D graphics engine, so things are not straightforward, you can't just pass any old userspace buffer to it. Short of some extreme bodgery that sets up a `pre_callback` function that reads from a file and writes directly over the the camera buffer, but that's difficult to recommend with a clear conscience...

**Comment by mwinter80** [comment]
Tricky indeed. Maybe a for now I just post it as a consideration for a future feature to allow offline development without a pi and/or a camera.

Thanks for your help!

**Comment by davidplowman** [comment]
Slightly against my better judgement, the following works:
```
import cv2
from picamera2 import Picamera2, MappedArray

size = (640, 360)
cap = cv2.VideoCapture("Downloads/Big_Buck_Bunny_360_10s_5MB.mp4")

picam2 = Picamera2()
config = picam2.create_preview_configuration({'size': size, 'format': 'XRGB8888'})
picam2.configure(config)

def pre_callback(request):
    ret, frame = cap.read()
    if ret:
        resized = cv2.resize(frame, size)
        with MappedArray(request, "main") as m:
            m.array[:, :, :3] = resized[:, :, :]

picam2.pre_callback = pre_callback
picam2.start(show_preview=True)
```
Of course, you _do_ still need a camera that runs, but if you're wanting a canned input stream for developement/testing, then it does kind of cover that.

**Comment by mwinter80** [comment]
Thanks.  I will try this out and let you know if I have any questions.

**Comment by mwinter80** [comment]
Unfortunately, I am on arch linux and picamera2 cannot find module libcamera. I tried to install libcamera from the aur package libcamera-git  but there are conflicts with pipewire, which I do not want to mess with. So I will have to put this on hold and continue developing directly on the pi. Thank you for your help anyway.
